mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h

```diff
@@ -0,0 +1,389 @@
+//===--- uArch.h ---------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+/// \file
+/// Base uArch definition for different architectures.
+///
+//
+//===----------------------------------------------------------------------===//
+#ifndef MLIR_DIALECT_XEGPU_UTILS_UARCH_H
+#define MLIR_DIALECT_XEGPU_UTILS_UARCH_H
+
+#include <any>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <mutex>
+#include <shared_mutex>
+#include <tuple>
+
+#include "mlir/IR/Types.h"
+
+namespace mlir {
+namespace xegpu {
+namespace uArch {
+
+// Data types we need for YAML to uArch translation
+struct Range {
+  int start;
+  int end;
+};
+
+// Restriction struct
+// This struct is used to represent a restriction on the uArch
+// The restriction is represented as a range of necessary parameters (template
+// arguments) and a lambda function (validate()) that takes the same number of
+// arguments as the number of template arguments The lambda function returns
+// true if the arguments satisfy the restriction The lambda function returns
+// false if the arguments do not satisfy the restriction
+
+// For example, a restriction that checks if the number of dimensions in a
+// std::vector<std::vector<uint32_t>> is 2 can be represented as:
+// std::vector<std::vector<uint32_t>> rt =
+// {{1, 32}, {2, 16}}; Restriction<std::vector<std::vector<uint32_t>>> r1(rt,
+// [](std::vector<std::vector<uint32_t>> t) { return t.size() == 2; });
+// r1.validate() will return true if the number of dimensions in the
+// std::vector<std::vector<uint32_t>> is 2 r1.validate() will return false if
+// the number of dimensions in the std::vector<std::vector<uint32_t>> is not 2
+
+// The primary purpose of Restriction struct is to provide a generic way to
+// represent restrictions on the uArch and to validate if the uArch satisfies
+// the restrictions
+template <typename... Args>
+struct Restriction {
+  std::tuple<Args...> data;
+  std::function<bool(Args...)> func;
+
+  Restriction(Args... args, std::function<bool(Args...)> f)
+      : data(args...), func(f) {}
+
+  bool validate() { return std::apply(func, data); }
+  std::any apply() { return std::apply(func, data); }
+};
+
+// Architecture HW component hierarchy to present thread, core, socket ...
+struct uArchHierarchyComponent {
+  std::string name = ""; // optional name of the hierarchy component
+  // no. of lower hierarchy component it contains, e.g., for PVC XeCore it
+  // contains 8 threads, so no_of_component=8
```

Comment:
Here in the comment that threads means 8 EU Threads? Or we 8 EUs (each run 4 threads under large register mode)?

Reply:
Yes.